presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -568,6 +561,74 @@ private void invokeNoMoreSplitsIfNecessary()
         return COMPLETED_FUTURE;
     }
 
+    private ListenableFuture<?> createHiveSymlinkSplits(ListenableFuture<?> lastResult, List<Path> targetPaths, InputFormat<?, ?> fileInputFormat, InternalHiveSplitFactory splitFactory)
+            throws IOException
+    {
+        for (Path targetPath : targetPaths) {
+            // The input should be in TextInputFormat.
+            // the splits must be generated using the file system for the target path
+            // get the configuration for the target path -- it may be a different hdfs instance
+            FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(hdfsContext, targetPath);
+            JobConf targetJob = toJobConf(targetFilesystem.getConf());
+            targetJob.setInputFormat(TextInputFormat.class);
+            Optional<Principal> principal = hdfsContext.getIdentity().getPrincipal();
+            if (principal.isPresent()) {
+                targetJob.set(MRConfig.FRAMEWORK_NAME, MRConfig.CLASSIC_FRAMEWORK_NAME);
+                targetJob.set(MRConfig.MASTER_USER_NAME, principal.get().getName());
+            }
+            if (fileInputFormat instanceof TextInputFormat) {
+                ((TextInputFormat) fileInputFormat).configure(targetJob);
+            }
+            FileInputFormat.setInputPaths(targetJob, targetPath);
+            InputSplit[] targetSplits = hdfsEnvironment.doAs(
+                    hdfsContext.getIdentity().getUser(),
+                    () -> fileInputFormat.getSplits(targetJob, 0));
+
+            lastResult = addSplitsToSource(targetSplits, splitFactory);
+            if (stopped) {
+                return COMPLETED_FUTURE;
+            }
+        }
+        return lastResult;
+    }
+
+    private Optional<Iterator<InternalHiveSplit>> buildManifestFileIterator(
+            Table table,
+            Path parent,
+            List<Path> paths,
+            DirectoryLister directoryLister,
+            NamenodeStats namenodeStats,
+            InternalHiveSplitFactory splitFactory,
+            boolean splittable)
+    {
+        Map<Path, LocatedFileStatus> fileStatuses = new HashMap();
+        try {
+            FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(hdfsContext, parent);
+            HiveFileIterator fileStatusIterator = new HiveFileIterator(table, parent, targetFilesystem, directoryLister, namenodeStats, IGNORED, true);
```

Comment:
nit: `HiveFileIterator` might skip some files (`io/prestosql/plugin/hive/util/HiveFileIterator.java:85`), but it should still work since we fallback to Hive native method when we cannot find matching `LocatedFileStatus` later on

Reply:
Good point. Will note that in the comment in `BackgroundHiveSplitLoader.java:619`